src/nvim/autocmd.c

```diff
@@ -1493,6 +1507,77 @@ void aucmd_restbuf(aco_save_T *aco)
   }
 }
 
+/// Schedules an autocommand event, to be executed at the next event-loop tick.
+///
+/// @param event event that occurred
+/// @param fname filename, NULL or empty means use actual file name
+/// @param fname_io filename to use for <afile> on cmdline,
+///                 NULL means use `fname`.
+/// @param group autocmd group ID or AUGROUP_ALL
+/// @param buf Buffer for <abuf>
+/// @param eap Ex command arguments
+void aucmd_defer(event_T event, char *fname, char *fname_io, int group, buf_T *buf, exarg_T *eap,
+                 Object *data)
+{
+  DeferredEvent *evdata = xmalloc(sizeof(DeferredEvent));
+  evdata->event = event;
+  evdata->fname = fname;
+  evdata->fname_io = fname_io;
+  evdata->group = group;
+  evdata->buf = buf->handle;
+  evdata->eap = eap;
+  if (data) {
+    evdata->data = xmalloc(sizeof(Object));
+    *evdata->data = copy_object(*data, NULL);
+  } else {
+    evdata->data = NULL;
+  }
+
+  multiqueue_put(deferred_autocmd_events, deferred_event, evdata);
+}
+
+/// Execute a deferred autocommand event
+static void deferred_event(void **argv)
+{
+  DeferredEvent *e = argv[0];
+  event_T event = e->event;
+  char *fname = e->fname;
+  char *fname_io = e->fname_io;
+  int group = e->group;
+  exarg_T *eap = e->eap;
+  Object *data = e->data;
+
+  Error err = ERROR_INIT;
+  buf_T *buf = find_buffer_by_handle(e->buf, &err);
+  if (buf) {
+    // set v:event with event data (if any)
```

Comment:
I found moving the `v:event` bit up avoided breaking any of the other tests. This shouldn't affect `nvim_exec_autocmds` although I guess `v:event` won't be set if `nvim_exec_autocmds("MarkSet", ...)` is called.

My main motivation for doing this copying is to make the testing in `test/functional/autocmd/markset_spec.lua` possible – I couldn't find any other autocommand tests that made use of the Lua-style `data`. Very much open to alternatives if we don't want this behavior.